Adds a unit test for Cargo Crates to prevent infinite credit oversights.

ArcaneMusic opened the pull request:
## About The Pull Request

I was kinda shocked that we didn't have something for this considering that it's an unspoken rule of cargo to check that a crate does not sell back for more than it's price and that the one comment saying to do so has been doing some heavy lifting for the last 12 years.

I'm a novice when it comes to unit testing specifically so hopefully the smoothing out that needs to be done should be fairly quick.

*Edit (2/22/25):*
The following changes were also made in order to allow for this unit test to work smoothly:

Exports now have a market define added to them, as the purpose of this unit test is to test exports that occur on the station, bought from supply and then sold back to supply.  As such, these market defines exclude exports specific to pirates (since the cargo shuttle cannot sell living mobs back, preventing you from making 10k per parrot crate.).

I've also added the `abstract` variable to some export datums, to signify that a given export is either variable, or not meant to be compared against the value of it's own container, such as with gas canister's base export value as their export datums are generated dynamically. (The subtypes are not abstract, however.)

The verb, `/mob/living/verb/tally_physical_credits()` has been changed to `/mob/living/proc/tally_physical_credits()`, because that's my B and does effect some economy back end but it's a one line fix so I just absent-mindedly fixed it here instead of atomizing it out. I can one-line it otherwise.

Mulebots now no longer runtime on spawn as they set their own to their own `get_turf` as opposed to pulling their `loc`.

A few supply packs have had their prices bumped up slightly to actually pass the test itself:

* `/datum/supply_pack/misc/candles_bulk`
* `/datum/supply_pack/security/armor`
* `/datum/supply_pack/security/helmets`
* `/datum/supply_pack/security/baton`

## Why It's Good For The Game

Prevents future infinite credit bugs that could have been missed by simply checking the sale value in game.

## Changelog

:cl:
fix: To prevent infinite sales issues, security helmets, armors, and batons packs now all cost 600 credits, up from 400.
fix: Candle packs now cost 400 credits, up from 300, and candles now sell for 12.25 cr each.
/:cl:

ArcaneMusic committed:
implements funny cargo crate sale unit testing

SmArtKar commented on code/modules/unit_tests/cargo_crate_sanity.dm:
```diff
@@ -0,0 +1,13 @@
+		var/obj/results = new_crate.generate(src)
+		var/datum/export_report/export_log = export_item_and_contents(results, apply_elastic = TRUE)
+		var/value = counterlist_sum(export_log.total_value)
+		TEST_ASSERT(value > CARGO_MINIMUM_COST, "Cargo crate [new_crate.name] had a sale value of [value], lower than [CARGO_MINIMUM_COST]")
```
Probably worth making sure that its also lower than its price, lest we end up with another fishing money fountain.

ArcaneMusic committed:
Merge remote-tracking branch 'upstream/master' into cargo-crate-sanity-unit-testing

ArcaneMusic committed:
Adds one more test against the crate's get_cost value.

SmArtKar commented on code/modules/unit_tests/cargo_crate_sanity.dm:
```diff
@@ -0,0 +1,14 @@
+/datum/unit_test/cargo_crate_sanity
+
+/datum/unit_test/cargo_crate_sanity/Run()
+	for(var/datum/supply_pack/new_crate in subtypesof(/datum/supply_pack))
```
This doesn't work, subtypesof gives you a list of, well, subtypes - not of objects. You need to init those yourself here, thus the name error and missing procs

SmArtKar commented on tgstation.dme:
```diff
@@ -6254,6 +6254,7 @@
 #include "code\modules\tutorials\tutorials\drop.dm"
 #include "code\modules\tutorials\tutorials\switch_hands.dm"
 #include "code\modules\unit_tests\_unit_tests.dm"
+#include "code\modules\unit_tests\cargo_crate_sanity.dm"
```
Shouldn't be doing that

ArcaneMusic committed:
Fixes

ArcaneMusic committed:
This is still busted, please excuse me

ArcaneMusic committed:
Cleans up the test assertion and failure usage

ArcaneMusic committed:
Take 2

ArcaneMusic committed:
Run this real quick for me?

ArcaneMusic committed:
Minor tweak please god test correctly

ArcaneMusic committed:
Adds abstract handling, balance candle export price lower, adds qdel handling on all contents

ArcaneMusic committed:
Lets try this I was cooking this on the drive down god I feel stupid that it took me this long but I did ask for help like, 5-6 times? Like damn, I feel stupid but I guess not that stupid, or perhaps more stupid because everyone must have thought that the fix was so easy that I surely must be bluffing about how simple the issue must be

ArcaneMusic committed:
Fixes a bunch of price issues, swaps to a get_turf to find clean testing floor

ArcaneMusic committed:
Merge remote-tracking branch 'upstream/master' into cargo-crate-sanity-unit-testing

ArcaneMusic committed:
Merge remote-tracking branch 'upstream/master' into cargo-crate-sanity-unit-testing

ArcaneMusic committed:
THIS IS THE ONE

ArcaneMusic committed:
attempts to stop hard delete issues.

ArcaneMusic committed:
Merge remote-tracking branch 'upstream/master' into cargo-crate-sanity-unit-testing

ArcaneMusic committed:
Oops :))

ArcaneMusic committed:
Still trying to fix this

ArcaneMusic committed:
comitting for testing purposes but it's basically dead air dw

LemonInTheDark committed:
Slightly unfucks space heater cell handling

Instead of HANGING A HARD REF TO A QDEL'D CELL the space heater now
releases it on Exited().
As a bonus, this means the heater will no longer drop a qdeleted cell
in to a locker on export due to a fucked up and stupid chain of logic
predicated on INSERTING THE CELL INTO ITS COMPONENT PARTS LIST ON DECON
WHY. I have left the sincode in because I am too tired to grasp why it's
here although I suspect it's just because whoever wrote it hates the
love of god

ArcaneMusic committed:
Merge pull request #9 from LemonInTheDark/use-to-fuel-this-feudal-monster

Slightly unfucks space heater cell handling

SmArtKar commented on code/modules/cargo/packs/security.dm:
```diff
@@ -20,7 +20,7 @@
 	name = "Armor Crate"
 	desc = "Three vests of well-rounded, decently-protective armor."
-	cost = CARGO_CRATE_VALUE * 2
+	cost = CARGO_CRATE_VALUE * 3
 	access_view = ACCESS_SECURITY
```
Add changelog for these 3

SmArtKar commented on code/modules/power/cell.dm:
```diff
@@ -103,6 +103,16 @@
 	custom_materials = list(/datum/material/glass=SMALL_MATERIAL_AMOUNT*0.6)
 	chargerate = STANDARD_CELL_RATE * 0.75
 
+/obj/item/stock_parts/power_store/cell/high/Initialize(mapload)
+	. = ..()
+
+/obj/item/stock_parts/power_store/cell/high/Destroy(force)
+	. = ..()
+
+/obj/item/stock_parts/power_store/cell/high/Move(atom/newloc, direct, glide_size_override, update_dir)
+	. = ..()
+	
+
```
???

ArcaneMusic committed:
Update cell.dm to rm empty overrides funni webedit edition

ArcaneMusic committed:
Merge branch 'master' into cargo-crate-sanity-unit-testing

ArcaneMusic committed:
Update secure.dm because I HATE WEBEDITS

ArcaneMusic committed:
I shrimply have decided that this is outside of my scope

ArcaneMusic committed:
Merge branch 'cargo-crate-sanity-unit-testing' of https://github.com/ArcaneMusic/TG-Station-Arcane-Remix into cargo-crate-sanity-unit-testing
